x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { APMPlugin, APMRouteHandlerResources } from '../..';
+import { listConfigurations } from '../settings/agent_configuration/list_configurations';
+import { setupRequest } from '../helpers/setup_request';
+import { APMPluginStartDependencies } from '../../types';
+import { ExternalCallback } from '../../../../fleet/server';
+
+export async function registerFleetPolicyCallbacks({
+  plugins,
+  ruleDataClient,
+  config,
+  logger,
+}: {
+  plugins: APMRouteHandlerResources['plugins'];
+  ruleDataClient: APMRouteHandlerResources['ruleDataClient'];
+  config: NonNullable<APMPlugin['currentConfig']>;
+  logger: NonNullable<APMPlugin['logger']>;
+}) {
+  if (!plugins.fleet) {
+    return;
+  }
+  const fleetPluginStart = await plugins.fleet.start();
+
+  registerAgentConfigExternalCallback({
+    fleetPluginStart,
+    callbackName: 'packagePolicyCreate',
+    plugins,
+    ruleDataClient,
+    config,
+    logger,
+  });
+  registerAgentConfigExternalCallback({
+    fleetPluginStart,
+    callbackName: 'packagePolicyUpdate',
+    plugins,
+    ruleDataClient,
+    config,
+    logger,
+  });
+}
+
+type ExternalCallbackParams = Parameters<ExternalCallback[1]>;
+type PackagePolicy = ExternalCallbackParams[0];
+type Context = ExternalCallbackParams[1];
+type Request = ExternalCallbackParams[2];
+
+function registerAgentConfigExternalCallback({
+  fleetPluginStart,
+  callbackName,
+  plugins,
+  ruleDataClient,
+  config,
+  logger,
+}: {
+  fleetPluginStart: NonNullable<APMPluginStartDependencies['fleet']>;
+  callbackName: ExternalCallback[0];
+  plugins: APMRouteHandlerResources['plugins'];
+  ruleDataClient: APMRouteHandlerResources['ruleDataClient'];
+  config: NonNullable<APMPlugin['currentConfig']>;
+  logger: NonNullable<APMPlugin['logger']>;
+}) {
+  const callbackFn: ExternalCallback[1] = async (
+    packagePolicy: PackagePolicy,
+    context: Context,
+    request: Request
+  ) => {
+    if (packagePolicy.package?.name !== 'apm') {
+      return packagePolicy;
+    }
+    const setup = await setupRequest({
+      context: context as any,
+      params: { query: { _inspect: false } },
+      core: null as any,
+      plugins,
+      request,
+      config,
+      logger,
+      ruleDataClient,
+    });
+    const configurations = await listConfigurations({ setup });
+    const agentConfigValue = configurations.map((configuration) => ({
+      service: configuration.service,
+      settings: configuration.settings,
+    }));
```

Comment:
```suggestion
      etag: configuration.etag,
      'agent.name': configuration.agent_name,
    }));
```

By passing the `etag` value down, apm-server will (later) be able to use it when recording which config has been applied by agents. The UI can then match this up to the config blocks it knows about.

We also need `agent.name` for restricting config for untrusted agents (RUM and Jaeger)